### ISSUE TITLE
Navigation bar not visible on highlighted articles in article list

### ISSUE THREAD

**Issue by emanruse** [newsboat/newsboat#1415]
```
newsboat 2.21.0 - https://newsboat.org/
Copyright (C) 2006-2015 Andreas Krennmair
Copyright (C) 2015-2020 Alexander Batischev
Copyright (C) 2006-2017 Newsbeuter contributors
Copyright (C) 2017-2020 Newsboat contributors

Newsboat is free software licensed under the MIT License. (Type `newsboat -vv' to see the full text.)
It bundles:
- JSON for Modern C++ library, licensed under the MIT License: https://github.com/nlohmann/json
- optional-lite library, licensed under the Boost Software License: https://github.com/martinmoene/optional-lite
- expected-lite library, licensed under the Boost Software License: https://github.com/martinmoene/expected-lite

newsboat 2.21.0
System: Linux 5.10.1-1-pae (i686)
Compiler: g++ 10.2.1 20200825 [revision c0746a1beb1ba073c7981eb09f55b3d993b32e5c]
ncurses: ncurses 6.2.20201205 (compiled with 6.2)
libcurl: libcurl/7.74.0 OpenSSL/1.1.1h-fips zlib/1.2.11 brotli/1.0.9 zstd/1.4.5 libidn2/2.3.0 libpsl/0.21.1 (+libidn2/2.3.0) libssh/0.9.5/openssl/zlib nghttp2/1.41.0 (compiled with 7.73.0)
SQLite: 3.34.0 (compiled with 3.33.0)
libxml2: compiled with 2.9.10
```

Steps to reproduce the issue:

1. Have some `highlight-article` lines such as:
```
highlight-article something red default bold
```
2. Navigate to a feed with articles containing that pattern

### The issue:

Whatever color or attribute style one chooses for highlighting, placing the cursor line on highlighted articles in article list (e.g. cyan default bold) does not show any visual hint that this is the selected line. The selection bar is partially visible only if there is empty trailing space to the right edge of the terminal (where there is no text).

**Comment by Minoru** [comment]
Reproduced with current master, e6742753aad8f0625ec2679a1e90afbc9113a6c3, and also r2.10 tag.

What is the expected behaviour, though? The line is drawn on top of the cursor, so the highlight style overrides cursor style. I think STFL lets us define a "focus" style which is applied only when the cursor is on that line, but what should that style contain?

**Comment by dennisschagt** [comment]
It looks like this happens because Newsboat uses the same style for the selected+highlighted item (`@style_%u_focus`) as for all other highlighted items (`"@style_%u_normal`):
https://github.com/newsboat/newsboat/blob/e6742753aad8f0625ec2679a1e90afbc9113a6c3/src/regexmanager.cpp#L355-L368
called from:
https://github.com/newsboat/newsboat/blob/e6742753aad8f0625ec2679a1e90afbc9113a6c3/src/itemlistformaction.cpp#L1411

I see two options for resolving this issue:
- Add a `highlight-article-focus` configuration (similar to `listfocus`/`listnormal` and `listfocus_unread`/`listnormal_unread` color options)
- Determine automatically which background color (and attributes?) to use based on the `listfocus` or `listfocus_unread` color configuration.

**Comment by Minoru** [comment]
Ah, cool, so there *is* an STFL style we can use for this!

Your second suggestion made me realize that focus could nullify the highlight. In other words, only non-focused lines would be subject to highlighting. But I guess this approahc might confuse people.

Also, your mention of colour options made me look at this from the styles point of view. We effectively have two ways to supply styles: via `color` and via `hightlight-article`. No wonder we have conflicts when those two styles clash! I see two ways to resolve that:

1. extend `highlight-article` to supply *two* styles, for "normal" and "focus" situations.

    This is similar to your first suggestion, but is IMHO nicer because it doesn't duplicate the filter expression. Compare:

    ```
    highlight-article "author = \"Joe\"" red green yellow
    highlight-article-focus "author = \"Joe\"" red green yellow underline
    ```

    to:

    ```
    highlight-article "author = \"Joe\"" normal: red green yellow focus: red green yellow underline
    ```

    Exact syntax subject to bikeshedding, of course. `normal:` and `focus:` are just tokens to make it easier for humans to parse this — Newsboat would technically be fine without them since colors and attributes have different names;

2. make the style in `highlight-article` optional, and add `listnormal_highlight` and `listfocus_highlight` elements to `color` setting.

    This option feels cleaner to me, because it means all colours are configured via the same command. However, that approach doesn't allow for styles to depend on the highlight rule. This is technically not a regression, but it makes the user decide what feature to give up: either highlight everything using a single style, or have many styles which don't play nicely with the cursor line.

> Determine automatically which background color (and attributes?) to use based on the listfocus or listfocus_unread color configuration.

Given how much trouble people have with writing colorschemes by hand (for everything, not just Newsboat), I doubt there's an automatic way to merge two styles in a pleasing way. For example, literally taking the background of a different style won't work if my highlight happens to use the same (or close) colour for the text itself. I'm not strictly against automation here, but I feel like we won't be able to automate it perfectly, and so will have to introduce a manual override, which is exactly the other solution we're discussing.

**Comment by emanruse** [comment]
> What is the expected behaviour, though?

The cursor line should be visible at all times (e.g. inverted or otherwise-customizable background).

> The line is drawn on top of the cursor, so the highlight style overrides cursor style.

Perhaps that needs to be corrected to match the expected behavior.

> I think STFL lets us define a "focus" style which is applied only when the cursor is on that line, but what should that style contain?

No idea what STFL stands for, so I am not sure what the question is. The general common sense expectation is to have a visual feedback.

**Comment by Minoru** [comment]
@emanruse STFL is a library we use to render the UI; it works on top of ncurses. The styles I'm talking about is the same styles that `highlight-article` and `color` use — colours (`red`, `black`, `color123` etc.) and attributes (`underline`, `bold` etc.).

**Comment by emanruse** [comment]
I see.
Well, my non-developer answer remains then :)

**Comment by marlun** [comment]
To me it would be OK that the focus color overrides the highlight-article color since you've specifically said how you want the focused item to look. The highlight-article is probably used mostly to find specific items in the list to actually highlight and open them?

**Comment by emanruse** [comment]
> To me it would be OK that the focus color overrides the highlight-article color since you've specifically said how you want the focused item to look.

What if that results in ambiguous color, e.g. the focus color being the same as the highlight-article color for the currently selected or for another item?

> The highlight-article is probably used mostly to find specific items in the list to actually highlight and open them?

Not sure what the question is. It is used for what the name says: to higlhight articles.

**Comment by marlun** [comment]
> What if that results in ambiguous color, e.g. the focus color being the same as the highlight-article color for the currently selected or for another item?

I'm not sure what you mean, the way I tried to describe it the highlight-article won't matter when the item is focused since the focus color will always be used. That seems more natural to me from what I've actually configured. And if you change the colors with highlight-article to the same as focus, then you've created your own mess :smile:

**Comment by emanruse** [comment]
> And if you change the colors with highlight-article to the same as focus, then you've created your own mess :smile:

No. That would be the result of what what you call "natural". Your suggestion allows the possibility to create a configuration leading to a visual (and hence also logical) collision. With such "natural" approach one must always remember not to use the color combination of highlight-article for anything else, which means additional complexity for the user and a limitation of the possible choices for the user. Complexity is the enemy of good UI/UX.

**Comment by marlun** [comment]
I don't agree on what's adding more complexity but arguing about it is probably not very helpful for the developers. We've both given them our thoughts and that's a good thing.

**Comment by emanruse** [comment]
Adding complexity (to code) is inevitable when a new feature is needed. That doesn't automatically mean that the final result will be more complex. The whole point of computing is to make human life easier.

That said, I agree it might be a good idea to make it simpler:

One constant color combination for focus. If you place the focus line on an article having the same highlight combination - blink the whole line (and blinking will not be a possible option for highlight-article).

What do you think?
